logstash-core/lib/logstash/util/safe_uri.rb

```diff
@@ -37,20 +38,132 @@ def inspect
   end
 
   def sanitized
-    return uri unless uri.password # nothing to sanitize here!
+    return uri unless password # nothing to sanitize here!
     
-    safe = uri.clone
-    safe.password = PASS_PLACEHOLDER
-    safe
+    user_info = user ? "#{user}:#{PASS_PLACEHOLDER}" : nil
+
+    # This syntax is required for jruby 1.7.x which doesn't reflect here correctly
+    # 9k does reflect here correctly. This can be removed when we are fully jruby 9k 
+    constructor = java.net.URI.java_class.constructor(java.lang.String, java.lang.String, java.lang.String, Java::int, java.lang.String, java.lang.String, java.lang.String)
+    java.net.URI.new(scheme, user_info, host, port || -1, path, query, fragment)
   end
 
   def ==(other)
     other.is_a?(::LogStash::Util::SafeURI) ? @uri == other.uri : false
   end
 
   def clone
-    cloned_uri = uri.clone
-    self.class.new(cloned_uri)
+    # No need to clone the URI, in java its immutable
+    self.class.new(uri)
+  end
+
+  def update(field, value)
+    new_scheme = scheme
+    new_user = user
+    new_password = password
+    new_host = host
+    new_port = port
+    new_path = path
+    new_query = query
+    new_fragment = fragment
+
+    case field 
+    when :scheme
+      new_scheme = field
+    when :user
+      new_user = field
+    when :password
+      new_password = field
+    when :host
+      new_host = field
+    when :port
+      new_port = field
+    when :path
+      new_path = field
+    when :query
+      new_query = field
+    when :fragment
+      new_fragment = field
+    end
+
+    user_info = new_user
+    if new_user && new_password
+      user_info += ":" + new_password
```

Comment:
@jordansissel its a good point! It gets tricky though. I'm trying to keep compatibility so code like: https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/808d73dd198255ef1439b65e33f5f93ff02c4706/lib/logstash/outputs/elasticsearch/http_client_builder.rb#L141 works.

The problem is I want this patch to behave like the old code used to, so that the ES output can stay compatible with old logstashes.

I don't like the lack of escaping, but I can't figure out a way out of it that isn't so complex that it isn't worth our time right this minute.